### PR TITLE
Add Go tutorial

### PR DESCRIPTION
**Pull request by infoverload**
## Summary of the changes / Why this is an improvement

Create 4th generate time series data tutorial (using Go)

## Checklist

 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

File: docs/generate-time-series/go.rst [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+        var err error
+        conn, err = pgx.Connect(context.Background(), "postgresql://crate@localhost:5432/doc")
+        if err != nil {
+            log.Fatalf("unable to connect to database: %v\n", err)
+        }
+	    defer conn.Close(context.Background())
+
+        conn.Exec(context.Background(), "CREATE TABLE iss (
+                                            timestamp TIMESTAMP GENERATED ALWAYS AS CURRENT_TIMESTAMP, 
+                                            position GEO_POINT)")
```
Comment (joemoe) [inline docs/generate-time-series/go.rst]: It appears to me that this code remains in the file all the time. 
should we use a `IF NOT EXISTS` upfront or drop the table before?
https://crate.io/docs/crate/reference/en/latest/sql/statements/create-table.html#if-not-exists

File: docs/generate-time-series/go.rst [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+            } else {
+                err = insertData(pos)
+                if err != nil {
+                    log.Fatalf("unable to insert data: %v\n", err)
+                }
+            }
+            fmt.Println("Sleeping for 5 seconds...")
+            time.Tick(time.Second * 5)
+	    }
+    }
```
Comment (joemoe) [inline docs/generate-time-series/go.rst]: there is no real _final_ version of the doc (of course same applies to other tutorials too), should we add this to a gist here?

File: docs/generate-time-series/go.rst [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+.. code-block:: console
+
+    $ go run main.go
+
+    Sleeping for 5 seconds...
+    Sleeping for 5 seconds...
+    Sleeping for 5 seconds...
+
+As the script runs, you should see the table filling up in the CrateDB Admin
+UI:
```
Comment (joemoe) [inline docs/generate-time-series/go.rst]: please add a link to the admin ui documentation or the section in the _getting started_ tutorial https://crate.io/docs/crate/tutorials/en/latest/getting-started/first-use/import.html#admin-ui

**Review by joemoe — COMMENTED** [review]
looking good, left some minor remarks.

File: docs/generate-time-series/go.rst [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+   :local:
+
+
+Prerequisites
+=============
+
+You must have CrateDB :ref:`installed and running <install-run>`.
+
+Make sure you're running an up-to-date version of `Go`_. We recommend 
+Go 1.11 or higher since we will be making use of `modules`_.
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]: can you reword this (and maybe other places?) to use the second person instead of the first person? cf. https://developers.google.com/style/person

File: docs/generate-time-series/go.rst [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+tooling since the `compilation`_ unit is the package and not a single line.
+
+To begin, create a project directory and navigate into it:
+
+.. code-block:: console
+
+    $ mkdir time-series-go
+    $ cd time-series-go
+
+Next, choose a module path and create a `go.mod` file that declares it:
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]:
```suggestion
Next, choose a module path and create a ``go.mod`` file that declares it:
```

literals require double backticks

Comment (nomicode) [reply docs/generate-time-series/go.rst]: (this applies in some other places in this file also)

Comment (nomicode) [reply docs/generate-time-series/go.rst]: unrelated to the above: what would you think about expanding this a little bit. it's not immediately clear to me what purpose the following command serves

File: docs/generate-time-series/go.rst [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+        "log"
+        "net/http"
+    )
+
+    func main() {
+
+    }
+
+Then, look at the JSON data that gets returned from going to the Open
+Notify API endpoint at `<http://api.open-notify.org/iss-now.json>`_. The
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]: do you want the reader to open this in the browser? or would it be better done on the command line? I think we should be explicit, either way :)

File: docs/generate-time-series/go.rst [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+   {
+        "message":"success",
+        "timestamp":1591703638,
+        "iss_position":{
+            "longitude":"84.9504",
+            "latitude":"41.6582"
+        }
+    }
+
+The longitude and latitude of the International Space Station changes
```
Comment (nomicode) [inline docs/generate-time-series/go.rst:119]: I would break this para up into two or more paras

File: docs/generate-time-series/go.rst [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+            Longitude string `json:"longitude"`
+            Latitude  string `json:"latitude"`
+        } `json:"iss_position"`
+    }
+
+Now, create a function that makes an HTTP GET request to the Open Notify API 
+endpoint and returns longitude and latitude as a `geo_point`_ declaration.
+
+The Go standard library comes with great support for HTTP client and server
+implementations in the `net/http`_ package. In this example you’ll use it to
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]: "In this example you’ll use it to..." is a little confusing from a grammar pov, imo. it sounds like you are making a prediction about the future wrt the reader's actions

I would stick to constructions like this:

- "Issue an HTTP GET request to the API endpoint" (when giving an instruction)
- "You can use this package to issue an HTTP GET request to the API endpoint" (when informing the reader about what they can do)

this is something I have been struggling with on the tutorial I am currently working on. how to use second person while also not making it sound like I'm making predictions about what the reader will do. I have found that "You can use X for Y" is a good solution -- that way, the tutorial takes the form of a series of instructions and pointers about what the reader can do with the tools/technology being demonstrated

does this make sense? 😅

File: docs/generate-time-series/go.rst [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+endpoint and returns longitude and latitude as a `geo_point`_ declaration.
+
+The Go standard library comes with great support for HTTP client and server
+implementations in the `net/http`_ package. In this example you’ll use it to 
+issue an HTTP GET request to the API endpoint. 
+
+Then you will read the response body and unmarshal the JSON into your defined 
+struct ``issInfo``. 
+
+Lastly, you will format the return string and return it.
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]: same as above

I would reword this:

"Lastly, you can format the return string and return it."

File: docs/generate-time-series/go.rst [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+        }
+
+        err = json.Unmarshal(responseData, &i)
+        if err != nil {
+            return "", fmt.Errorf("unable to unmarshal response body: %v", err)
+        }
+
+        s := fmt.Sprintf("(%s, %s)", i.IssPosition.Longitude, i.IssPosition.Latitude)
+        return s, nil
+    }
```
Comment (nomicode) [inline docs/generate-time-series/go.rst:167]: I would introduce this code block with something like, "Here's a function that [...]:"

then afterward, I would consider adding some text to explain what is happening, step by step. e.g., "Above, the `getISSPosition` function does the following:" -- you might want to take the explanation above this function and move it here to serve this purpose

then, you can list out a bunch things, explaining the function as you go

---

to help illustrate my point, here's an excerpt from the tutorial I am working on at the moment that appears directly after a Python code block:

> Above, the ``plot()`` function:
>
> * Generates a [figure]() that measures 12 × 6 (inches)
> * Plots ``data`` as a [scatter]() diagram (distance over time)
> * Sets the [axes]() labels and title
> * Sets up the x-axis to [handle datetimes]()
> * Configures major [tick locations]() every [hour]()
> * Configures major [tick formatting]() with a [time string]() (``%H:00``)
> * Forces Matplotlib to plot the whole data set, including null ``time`` values, by manually setting the [limits of the x-axis]() (which are trimmed by default)
> * Activates x-axis tick label [auto-formatting]() (rotates them for improved readability)

File: docs/generate-time-series/go.rst [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+
+In your main function, call your ``getISSPosition()`` function and print out the result:
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]: would use "the" instead of "your" here. "you", "your" for talking about things the reader should do, has done, etc, but the definite article ("the") for data/code we are sharing with the reader

File: docs/generate-time-series/go.rst [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+
+.. code-block:: console
+
+    $ go run main.go
+
+The result should contain your geo_point string:
+
+.. code-block:: js
+
+    (104.7298, 5.0335)
```
Comment (nomicode) [inline docs/generate-time-series/go.rst:202]: I would add a bit that mentions how you can run this multiple times to get the new position of the ISS each time (like how you do later when inserting)

File: docs/generate-time-series/go.rst [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+        }
+	    defer conn.Close(context.Background())
+
+        conn.Exec(context.Background(), "CREATE TABLE iss (
+                                            timestamp TIMESTAMP GENERATED ALWAYS AS CURRENT_TIMESTAMP, 
+                                            position GEO_POINT)")
+    }
+
+
+Save all your changes and run the new code in the command line:
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]: I would drop "in the command line" here and elsewhere. I believe it's implied

File: docs/generate-time-series/go.rst [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+Save all your changes and run the code in the command line:
+
+.. code-block:: console
+
+    $ go run main.go
+
+Press the up arrow on your keyboard and hit *Enter* to run the same command a
+few more times.
+
+When you're done, you can `SELECT`_ that data back out of CrateDB by executing
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]: what do you think about including an example of the output here? perhaps you could reformat the table that is displayed using RST tables

[edit: on second thoughts, given that we already include a screenshot of the admin UI elsewhere in these tutorials, perhaps a screenshot would be better than a table)

this also makes me think that the CREATE TABLE command would be easier to run from the Admin UI

File: docs/generate-time-series/go.rst [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+In your file ``main.go``, create a function that encapsulates data insertion:
+
+.. code-block:: js
+
+    func insertData(position string) error {
+        _, err := conn.Exec(context.Background(), "INSERT INTO iss (position) VALUES ($1)", position)
+        return err
+    }
+
+Then in the ``main`` function of your script, create an infinite loop that
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]: I would break up this para

File: docs/generate-time-series/go.rst [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+
+Then in the ``main`` function of your script, create an infinite loop that 
+gets the latest ISS position and inserts the data into the database.  You 
+also have some basic error handling, in case either the API query or the
+CrateDB operation fails.  The script sleeps for 5 seconds after each sample. 
+Accordingly, the time series data will have a *resolution* of 5 seconds. If 
+you wish to change this resolution, you may want to configure your script 
+differently.
+
+Don't forget to import the `time`_ package.
```
Comment (nomicode) [inline docs/generate-time-series/go.rst]: like I mentioned in a previous comment, I am wondering if this would be easier to understand from the reader's pov if the code block came first and the explanation came after

File: docs/generate-time-series/go.rst [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+	    }
+    }
+
+Run the script from the command line:
+
+.. code-block:: console
+
+    $ go run main.go
+
+    Sleeping for 5 seconds...
```
Comment (nomicode) [inline docs/generate-time-series/go.rst:365]: can we make it so the output is the same as the other scripts? (i.e., add "CONNECT OK", "INSERT OK")

**Review by nomicode — COMMENTED** [review]
yay, thank you!! left some comments!!

there are a few trivial copyedits I would make that I haven't commented on for this review. I figure that once all the non-trivial stuff has been discussed I can just add a copyedit commit to your PR for your review

File: docs/generate-time-series/go.rst (hunk 14 above)
Comment (infoverload) [reply docs/generate-time-series/go.rst]: Yeah I wasn't really sure which to go with (writing logic in the script or using the Admin Console).  In this case, since I am trying to create 1 script, I thought it would make more sense to not include a function that just selects back out all the data and displays it on stdout.  With CREATE TABLE, if I change it to CREATE TABLE [ IF NOT EXISTS ], then maybe it's a good example to show in the code snippet?

Comment (infoverload) [reply docs/generate-time-series/go.rst]: And given that we already include a screenshot of the admin UI near the end of the tutorial, would it be redundant to include a very similar one again...or would it be consistent?

Comment (nomicode) [reply docs/generate-time-series/go.rst]: yes, that makes sense. okay! it's fine as is imo :)

**Comment by nomicode** [comment]
looks good, @infoverload! wanna run this through Grammarly?

**Comment by nomicode** [comment]
if https://github.com/crate/crate-tutorials/pull/7 is merged before this PR, this PR should be updated to match the new phrasing used across tutorials

**Review by nomicode — APPROVED** [review]
